backend/clickhouse/traces.go

```diff
@@ -316,63 +318,79 @@ func (client *Client) ReadTrace(ctx context.Context, projectID int, traceID stri
 	return traces, rows.Err()
 }
 
-func (client *Client) ReadTracesMetrics(ctx context.Context, projectID int, params modelInputs.QueryInput, column modelInputs.TracesMetricColumn, metricTypes []modelInputs.MetricAggregator, groupBy []string, nBuckets int) (*modelInputs.TracesMetrics, error) {
+func getFnStr(aggregator modelInputs.MetricAggregator, column string, useSampling bool) string {
+	switch aggregator {
+	case modelInputs.MetricAggregatorCount:
+		if useSampling {
+			return "round(count() * any(_sample_factor))"
+		} else {
+			return "toFloat64(count())"
+		}
+	case modelInputs.MetricAggregatorCountDistinctKey:
+		if useSampling {
+			return fmt.Sprintf("round(count(distinct TraceAttributes['%s']) * any(_sample_factor))", highlight.TraceKeyAttribute)
+		} else {
+			return fmt.Sprintf("toFloat64(count(distinct TraceAttributes['%s']))", highlight.TraceKeyAttribute)
+		}
+	case modelInputs.MetricAggregatorMin:
+		return fmt.Sprintf("toFloat64(min(%s))", column)
+	case modelInputs.MetricAggregatorAvg:
+		return fmt.Sprintf("avg(%s)", column)
+	case modelInputs.MetricAggregatorP50:
+		return fmt.Sprintf("quantile(.5)(%s)", column)
+	case modelInputs.MetricAggregatorP90:
+		return fmt.Sprintf("quantile(.9)(%s)", column)
+	case modelInputs.MetricAggregatorP95:
+		return fmt.Sprintf("quantile(.95)(%s)", column)
+	case modelInputs.MetricAggregatorP99:
+		return fmt.Sprintf("quantile(.99)(%s)", column)
+	case modelInputs.MetricAggregatorMax:
+		return fmt.Sprintf("toFloat64(max(%s))", column)
+	case modelInputs.MetricAggregatorSum:
+		if useSampling {
+			return fmt.Sprintf("sum(%s) * any(_sample_factor)", column)
+		} else {
+			return fmt.Sprintf("sum(%s)", column)
+		}
+	}
+	return ""
+}
+
+func (client *Client) ReadTracesMetrics(ctx context.Context, projectID int, params modelInputs.QueryInput, column string, metricTypes []modelInputs.MetricAggregator, groupBy []string, nBuckets int, bucketBy string, limit *int, limitAggregator *modelInputs.MetricAggregator, limitColumn *string) (*modelInputs.TracesMetrics, error) {
 	if len(metricTypes) == 0 {
 		return nil, e.New("no metric types provided")
 	}
 
+	if bucketBy == modelInputs.TracesMetricBucketByNone.String() {
+		nBuckets = 1
+	}
+
 	startTimestamp := uint64(params.DateRange.StartDate.Unix())
 	endTimestamp := uint64(params.DateRange.EndDate.Unix())
 	// always sample - use a comparison here to trick the compiler into not complaining about unused branches
 	useSampling := params.DateRange.EndDate.Sub(params.DateRange.StartDate) >= 0
 
-	metricColName := "Duration"
+	var metricColName string
+	if col, found := traceKeysToColumns[modelInputs.ReservedTraceKey(column)]; found {
+		metricColName = col
+	} else {
+		metricColName = fmt.Sprintf("toFloat64OrNull(TraceAttributes['%s'])", column)
+	}
+
```

Comment:
is this susceptible to sql injection if `column` is user-provided, since this now allows column names other than the reserved one

Reply:
ah yes, thanks for spotting! fixing